build-tools/Xamarin.Android.Tools.BootstrapTasks/Xamarin.Android.Tools.BootstrapTasks/RunParallelCmds.cs

```diff
@@ -0,0 +1,65 @@
+using System;
+using System.Diagnostics;
+
+using Microsoft.Build.Framework;
+using Microsoft.Build.Utilities;
+
+using Tasks = System.Threading.Tasks;
+
+namespace Xamarin.Android.Tools.BootstrapTasks
+{
+	public class RunParallelCmds : Task
+	{
+		[Required]
+		public ITaskItem[] Commands { get; set; }
+
+		public string ManagedRuntime { get; set; }
+
+		public string ManagedRuntimeArguments { get; set; }
+
+		public override bool Execute ()
+		{
+			bool success = true;
+
+			Tasks.Parallel.ForEach (Commands, new Tasks.ParallelOptions () { MaxDegreeOfParallelism = Environment.ProcessorCount },
+				cmd => {
+					var useManagedRuntime = !string.IsNullOrEmpty (ManagedRuntime);
+					var argumentsBeginning = useManagedRuntime ? ManagedRuntimeArguments + " " : "";
+					var procStartInfo = new ProcessStartInfo () {
+						FileName =  useManagedRuntime ? ManagedRuntime : cmd.GetMetadata ("Command"),
+						Arguments = $"{argumentsBeginning}{cmd.GetMetadata ("Arguments")}",
+						CreateNoWindow = true,
+						UseShellExecute = false,
+						RedirectStandardOutput = true,
+						RedirectStandardError = true,
+					};
+
+					Log.LogMessage (MessageImportance.Normal, $"Starting Command: {cmd.GetMetadata ("Command")} Arguments: {cmd.GetMetadata ("Arguments")}");
+
+					try {
+						using (var proc = Process.Start (procStartInfo)) {
+							proc.WaitForExit ();
+
+							if (proc.ExitCode != 0) {
+								Log.LogMessage (MessageImportance.High, $"Non-zero exit code: {proc.ExitCode}  Error output: {proc.StandardError.ReadToEnd ()}");
```

Comment:
You should always read `proc.StandardError`, not only when `ExitCode != 0`. There might be output on `stderr ` even if `ExitCode == 0` and then the task may hang.

Reply:
Indeed, good catch.

Reply:
I think I might need to redo the output reading as it might hang when one of the buffers get full.

It works here so far because there's no much output from these tools.

Reply:
Yep, but if we ran, say, `git`, then it can and will produce its debug output on `stderr` (not saying that we will run git in parallel, but we might run into something similar :D)

Reply:
@radekdoulik maybe you can take advantage of https://github.com/xamarin/xamarin-android/blob/master/build-tools/xaprepare/xaprepare/Application/ProcessRunner.cs?

Reply:
So I will add handlers for `ErrorDataReceived` and `OutputDataReceived ` events, call `BeginOutputReadLine` and `BeginErrorReadLine` before calling `WaitForExit`.